### PR TITLE
Some more symbol docs modified

### PR DESCRIPTION
**Pull request by Roshrini**
Modified some of symbol docs like: __iter__, __call__, _compose, bind, simple_bind, eval, var, save, __neg__, attr, list_attr, attr_dict.

@mli @zackchase @madjam @nswamy @indhub @jiajiechen Please take a look.

### CODE REVIEW

File: python/mxnet/symbol.py [hunk 1]
```diff
@@ -51,7 +51,25 @@ def __repr__(self):
                             'Grouped' if name is None else name)
 
     def __iter__(self):
-        """Returns all outputs in a list"""
+        """Returns a generator object of symbol.
+        You can loop through the returned object to get outputs.
```
Comment (madjam) [inline python/mxnet/symbol.py]: You -> One

File: python/mxnet/symbol.py [hunk 2]
```diff
@@ -136,7 +154,23 @@ def __pow__(self, other):
             raise TypeError('type %s not supported' % str(type(other)))
 
     def __neg__(self):
-        """x.__neg__(y) <=> -x """
+        """x.__neg__() <=> -x
+        This function applies negative to the input, element-wise.
```
Comment (madjam) [inline python/mxnet/symbol.py]: Suggest changing to: "Numerical negative, element-wise."
This is how numpy describes it.

File: python/mxnet/symbol.py [hunk 3]
```diff
@@ -250,37 +284,65 @@ def __setstate__(self, state):
             self.handle = None
 
     def __call__(self, *args, **kwargs):
-        """Compose symbol on inputs.
+        """Composes symbol on inputs.
```
Comment (madjam) [inline python/mxnet/symbol.py]: on -> using

File: python/mxnet/symbol.py [hunk 4]
```diff
@@ -250,37 +284,65 @@ def __setstate__(self, state):
...
-        the resulting symbol
+            The resulting symbol.
         """
         s = self.__copy__()
         s._compose(*args, **kwargs)
         return s
 
     def _compose(self, *args, **kwargs):
-        """Compose symbol on inputs.
+        """Composes symbol on inputs.
```
Comment (madjam) [inline python/mxnet/symbol.py]: on -> using?

File: python/mxnet/symbol.py [hunk 5]
```diff
@@ -368,7 +430,14 @@ def name(self):
             return None
 
     def attr(self, key):
-        """Gets attribute string from the symbol. This function only works for non-grouped symbols.
+        """Returns an attribute string for corresponding input key from the symbol.
```
Comment (madjam) [inline python/mxnet/symbol.py]: an -> the

File: python/mxnet/symbol.py [hunk 6]
```diff
@@ -898,17 +981,19 @@ def save(self, fname):
         """Saves symbol to a file.
 
         You can also use pickle to do the job if you only work on python.
-        The advantage of load/save is the file is language agnostic.
-        This means the file saved using save can be loaded by other language binding of mxnet.
-        You also get the benefit being able to directly load/save from cloud storage(S3, HDFS)
+        The advantage of `load`/`save` functions is that the file is language agnostic.
```
Comment (madjam) [inline python/mxnet/symbol.py]: "is that the file is language..." -> "is that the file contents are language..."

File: python/mxnet/symbol.py [hunk 7]
```diff
@@ -898,17 +981,19 @@ def save(self, fname):
...
+        This means that the file saved using `save` function can be loaded by other
```
Comment (madjam) [inline python/mxnet/symbol.py]: Suggest changing this to:
This means the model saved by one language binding can be loaded by a different language binding of MXNet.

File: python/mxnet/symbol.py [hunk 8]
```diff
@@ -990,12 +1075,30 @@ def simple_bind(self, ctx,
...
                     **kwargs):
-        """Bind current symbol to get an executor, allocate all the ndarrays needed.
+        """Binds current symbol to get an executor, allocate all the arguments needed.
         Allows specifying data types.
 
-        This function will ask user to pass in an `NDArray` of position
-        they like to bind to, and it will automatically allocate the ndarray
-        for arguments and auxiliary states that user did not specify explicitly.
+        This function simplifies the binding procedure. You need to specify only input data shapes.
+        The function allocates the arguments and auxiliary states that you did not specify
```
Comment (madjam) [inline python/mxnet/symbol.py]: Suggest rewording as:
"Before binding the executor, the function allocates arguments and auxiliary states that were not explicitly specified."

File: python/mxnet/symbol.py [hunk 9]
```diff
@@ -1068,7 +1172,24 @@ def simple_bind(self, ctx,
 
     def bind(self, ctx, args, args_grad=None, grad_req='write',
              aux_states=None, group2ctx=None, shared_exec=None):
-        """Bind current symbol to get an executor.
+        """Binds the current symbol to get an executor.
```
Comment (madjam) [inline python/mxnet/symbol.py]: How about?
"Binds the current symbol to an executor and returns it."

File: python/mxnet/symbol.py [hunk 10]
```diff
@@ -1068,7 +1172,24 @@ def simple_bind(self, ctx,
...
+
+        We first declare the computation and then bind it with the data to run.
```
Comment (madjam) [inline python/mxnet/symbol.py]: it -> to

File: python/mxnet/symbol.py [hunk 11]
```diff
@@ -1068,7 +1172,24 @@ def simple_bind(self, ctx,
...
+        This function returns an executor which provides method `forward()` for evaluation
```
Comment (madjam) [inline python/mxnet/symbol.py]: How about:

This function returns an executor which provides a `forward()` method for evaluation and a `outputs()` method to get all the results.

File: python/mxnet/symbol.py [hunk 12]
```diff
@@ -1078,38 +1199,40 @@ def bind(self, ctx, args, args_grad=None, grad_req='write',
...
               to the corresponding NDArray.
-            - When the type is dict of str to `NDArray`, users only need to provide the dict
-              for needed argument gradient.
+            - When the type is a dict of str to `NDArray`, you only need to provide the dict
+              for required argument gradient.
               Only the specified argument gradient will be calculated.
 
         grad_req : {'write', 'add', 'null'}, or list of str or dict of str to str, optional
-            Specifies how we should update the gradient to the args_grad.
+            To Specify how we should update the gradient to the `args_grad`.
```
Comment (madjam) [inline python/mxnet/symbol.py]: Lower case s in Specify

File: python/mxnet/symbol.py [hunk 13]
```diff
@@ -1078,38 +1199,40 @@ def bind(self, ctx, args, args_grad=None, grad_req='write',
...
 
-            - If type is list of `NDArray`, the position is in the same order of list_arguments.
-            - If type is dict of str to `NDArray`, then it maps the name of arguments
+            - If the input type is a list of `NDArray`, the order should be same as the order
+              of `list_arguments()`.
+            - If the input type is a dict of str to `NDArray`, then it maps the name of arguments
               to the corresponding NDArray.
-            - When the type is dict of str to `NDArray`, users only need to provide the dict
-              for needed argument gradient.
+            - When the type is a dict of str to `NDArray`, you only need to provide the dict
```
Comment (madjam) [inline python/mxnet/symbol.py]: you -> one

File: python/mxnet/symbol.py (hunk 12 above)
Comment (madjam) [inline python/mxnet/symbol.py]: Same here. Lower case s.

File: python/mxnet/symbol.py [hunk 14]
```diff
@@ -1078,38 +1199,40 @@ def bind(self, ctx, args, args_grad=None, grad_req='write',
...
+            To Specify how we should update the gradient to the `args_grad`.
 
             - 'write' means everytime gradient is write to specified args_grad `NDArray`.
             - 'add' means everytime gradient is add to the specified NDArray.
             - 'null' means no action is taken, the gradient may not be calculated.
 
         aux_states : list of `NDArray`, or dict of str to `NDArray`, optional
-            Input auxiliary states to the symbol, only need to specify when
-            list_auxiliary_states is not empty.
+            Input auxiliary states to the symbol, only need to specify when the output of
```
Comment (madjam) [inline python/mxnet/symbol.py]: "need to specify" -> needed

File: python/mxnet/symbol.py [hunk 15]
```diff
@@ -1122,12 +1245,12 @@ def bind(self, ctx, args, args_grad=None, grad_req='write',
...
         executor : Executor
-            The generated Executor
+            The generated executor
 
         Notes
         -----
-        Auxiliary states are special states of symbols that do not correspond
-        to an argument, and do not have gradient. But still be useful
+        Auxiliary states are the special states of symbols that do not correspond
+        to an argument, and do not have gradient but still are useful
```
Comment (madjam) [inline python/mxnet/symbol.py]: still are -> are still

File: python/mxnet/symbol.py [hunk 16]
```diff
@@ -1232,61 +1355,78 @@ def grad(self, wrt):
...
+    <Symbol v>
 
     Parameters
     ----------
     name : str
-        Name of the variable.
-    attr : dict of string -> string
-        Additional attributes to set on the variable.
+        Variable name.
+    attr : dict of strings
```
Comment (madjam) [inline python/mxnet/symbol.py]: dict -> Dict

File: python/mxnet/symbol.py [hunk 17]
```diff
@@ -1232,61 +1355,78 @@ def grad(self, wrt):
...
-    """Create a symbolic variable with specified name.
+    """Creates a symbolic variable with specified name.
+    Equivalent to mx.sym.Variable.
+
+    Example usage:
+    ----------
+    >>> data = mx.sym.Variable('data', attr={'a': 'b'})
+    >>> data
+    <Symbol data>
+    >>> data = mx.sym.Variable('data', shape = (2,3), __shape__=(9,0))
```
Comment (piiswrong) [inline python/mxnet/symbol.py]: dont set __shape__

Comment (Roshrini) [inline python/mxnet/symbol.py]: @piiswrong I added shape to show that if it is specified in kwargs(__shape__), it will overwrite shape from (2,3) to (9,0).
Should I remove it?

**Comment by Roshrini** [comment]
@madjam Did the mentioned changes.

File: python/mxnet/symbol.py [hunk 18]
```diff
@@ -51,11 +51,31 @@ def __repr__(self):
                             'Grouped' if name is None else name)
 
     def __iter__(self):
-        """Returns all outputs in a list"""
+        """Returns a generator object of symbol.
+
+        One can loop through the returned object to get outputs.
```
Comment (zackchase) [inline python/mxnet/symbol.py]: Is the object a list? If so, let's call it a list. Sounds a bit strange to say "loop through the object"

File: python/mxnet/symbol.py [hunk 19]
```diff
@@ -80,14 +101,26 @@ def __sub__(self, other):
             raise TypeError('type %s not supported' % str(type(other)))
 
     def __rsub__(self, other):
-        """x.__rsub__(y) <=> y-x """
+        """x.__rsub__(y) <=> y-x
+
+        Only NDArray is supported for now.
```
Comment (zackchase) [inline python/mxnet/symbol.py]: place NDArray in backticks

File: python/mxnet/symbol.py [hunk 20]
```diff
@@ -112,7 +146,18 @@ def __div__(self, other):
             raise TypeError('type %s not supported' % str(type(other)))
 
     def __rdiv__(self, other):
-        """x.__rdiv__(y) <=> y/x """
+        """x.__rdiv__(y) <=> y/x
+
+        Only NDArray is supported for now.
```
Comment (zackchase) [inline python/mxnet/symbol.py]: ""

File: python/mxnet/symbol.py [hunk 21]
```diff
@@ -250,37 +319,67 @@ def __setstate__(self, state):
...
         return s
 
     def _compose(self, *args, **kwargs):
-        """Compose symbol on inputs.
+        """Composes symbol using inputs.
 
-        This call mutates the current symbol.
+        x._compose(y, z) <=> x(y,z)
+
+        This function call mutates the current symbol.
```
Comment (zackchase) [inline python/mxnet/symbol.py]: Can we just say "function" instead of "function call"?

File: python/mxnet/symbol.py [hunk 22]
```diff
@@ -378,7 +485,7 @@ def attr(self, key):
         Returns
         -------
         value : str
-            The desired attribute value, returns None if attribute does not exist.
+            The desired attribute value, returns `None` if the attribute does not exist.
```
Comment (zackchase) [inline python/mxnet/symbol.py]: Let's put None in double backticks per standard we've established

File: python/mxnet/symbol.py [hunk 23]
```diff
@@ -898,17 +1022,19 @@ def save(self, fname):
         """Saves symbol to a file.
 
         You can also use pickle to do the job if you only work on python.
-        The advantage of load/save is the file is language agnostic.
-        This means the file saved using save can be loaded by other language binding of mxnet.
-        You also get the benefit being able to directly load/save from cloud storage(S3, HDFS)
+        The advantage of `load`/`save` functions is that the file contents are language agnostic.
+        This means the model saved by one language binding can be loaded by a different
+        language binding of MXNet.
```
Comment (zackchase) [inline python/mxnet/symbol.py]: MXNet in backticks

File: python/mxnet/symbol.py [hunk 24]
```diff
@@ -1004,24 +1147,25 @@ def simple_bind(self, ctx,
 
         grad_req: string
             {'write', 'add', 'null'}, or list of str or dict of str to str, optional
-            Specifies how we should update the gradient to the args_grad.
-            - 'write' means everytime gradient is write to specified args_grad NDArray.
+            To specify how we should update the gradient to the `args_grad`.
+
+            - 'write' means everytime gradient is write to specified `args_grad` NDArray.
```
Comment (zackchase) [inline python/mxnet/symbol.py]: This sentence needs revision.

"everytime" is not a word - always write "every time". 
also "is write" -> "is written"

**Comment by Roshrini** [comment]
@zackchase Done the suggested changes.

File: python/mxnet/symbol.py (hunk 17 above)
Comment (piiswrong) [inline python/mxnet/symbol.py]: yes, please remove `__shape__` its for internal use only.

File: python/mxnet/symbol.py [hunk 25]
```diff
@@ -51,11 +51,31 @@ def __repr__(self):
...
+        ----------
+        >>> a = mx.sym.Variable('a')
+        >>> b = mx.sym.Variable('b')
+        >>> c = a+b
+        >>> d = mx.sym.Variable('d')
+        >>> e = d+c
+        >>> out = e.get_children()
+        >>> out
+        <Symbol Grouped>
+        >>> for i in out.__iter__():
```
Comment (piiswrong) [inline python/mxnet/symbol.py]: do you need to call __iter__() explicitly?

Comment (Roshrini) [inline python/mxnet/symbol.py]: No. you can choose not to call it explicitly. I called it to show the example call to the function. I will remove it
